File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Base Condition
        if (head == None or head.next == None):
            return head
        # iniitalisation
        prevNode = head
        currNode = head.next
        nextNode = currNode.next
        # reverse the links
        while (currNode != None):
            nextNode = currNode.next
            currNode.next = prevNode
            prevNode = currNode
            currNode = nextNode
            

        head.next = None
        return prevNode 

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        //Base Condition
        if (head == NULL || head->next == NULL) {
            return false;
        }

        //Atleast 2 nodes are there
        ListNode* curNode = head;
        ListNode* fastNode = head->next;
        int counter = 0;

        while (fastNode->next != NULL && fastNode->next->next !=NULL) {
            if (curNode == fastNode) {
                return true;
            }
            curNode = curNode->next;
            fastNode = fastNode->next->next;
        }

        return false;
    }
};
